zerver/views/message_summary.py

```diff
@@ -0,0 +1,131 @@
+    # Rolling Summaries are summaries that are generated for a fixed number of messages
+    # at a time, and then the next summary is generated for the next fixed number of
+    # messages with the previous summary as the starting point. This way, we can
+    # generate summaries for new messages in a single pass.
+    # TODO: Come up with a plan to store these summaries in the database.
+
+    messages_response = get_messages_backend(
+        request,
+        user_profile,
+        narrow=narrow,
+        anchor_val="newest",
+        num_before=MAX_MESSAGES_SUMMARIZED,
+        client_gravatar=False,
+        apply_markdown=False,
+    )
+    zulip_messages = json.loads(messages_response.content).get("messages", [])
+    if len(zulip_messages) == 0:
+        return json_success(request, {"summary": "No messages in conversation to summarize"})
+
+    # XXX: Translate input and output text to English?
+    model = settings.TOPIC_SUMMARIZATION_MODEL
+    litellm_params: dict[str, Any] = {}
+    if model.startswith("huggingface"):
+        assert settings.HUGGINGFACE_API_KEY is not None
+        litellm_params["api_key"] = settings.HUGGINGFACE_API_KEY
+    else:
+        assert model.startswith("bedrock")
+        litellm_params["aws_access_key_id"] = settings.AWS_ACCESS_KEY_ID
+        litellm_params["aws_secret_access_key"] = settings.AWS_SECRET_ACCESS_KEY
+        litellm_params["aws_region_name"] = settings.AWS_REGION_NAME
+
+    conversation_length = len(zulip_messages)
+    max_summary_length = get_max_summary_length(conversation_length)
+    intro = "The following is a chat conversation in the Zulip team chat app."
+    topic: str | None = None
+    channel: str | None = None
+    if narrow and len(narrow) == 2:
+        for term in narrow:
+            assert not term.negated
+            if term.operator == "channel":
+                channel = term.operand
+            if term.operator == "topic":
+                topic = term.operand
+    if channel:
+        intro += f" channel: {channel}"
+    if topic:
+        intro += f", topic: {topic}"
+
+    formatted_conversation = format_conversation(zulip_messages)
+    prompt = (
+        f"Succinctly summarize this conversation based only on the information provided, "
+        f"in up to {max_summary_length} sentences, for someone who is familiar with the context. "
+        f"Mention key conclusions and actions, if any. Refer to specific people as appropriate. "
+        f"Don't use an intro phrase."
+    )
+    messages = [
+        make_message(intro, "system"),
+        make_message(formatted_conversation),
+        make_message(prompt),
+    ]
+
+    # Token counter is recommended by LiteLLM but mypy says it's not explicitly exported.
+    # https://docs.litellm.ai/docs/completion/token_usage#3-token_counter
+    input_tokens = litellm.token_counter(model=model, messages=messages)  # type: ignore[attr-defined] # Explained above
```

Comment:
What does this function actually do? I worry it may do an API call to ask the model to parse the input into tokens, and thus result in a charge for it.

Reply:
token_counter: This returns the number of tokens for a given input - it uses the tokenizer based on the model, and defaults to tiktoken if no model-specific tokenizer is available. 

https://github.com/BerriAI/litellm/blob/main/litellm/utils.py#L1535

I don't see any netowrk calls when running the code, so it is purely offline. 

We could completely skip checking `input_tokens` here based on how we want to bill the user.